srv/src/main/java/my/bookshop/handlers/CatalogServiceHandler.java

```diff
@@ -2,56 +2,223 @@
+import com.sap.cds.services.cds.CdsReadEventContext;
 import com.sap.cds.services.cds.CdsService;
 import com.sap.cds.services.handler.EventHandler;
 import com.sap.cds.services.handler.annotations.After;
+import com.sap.cds.services.handler.annotations.Before;
+import com.sap.cds.services.handler.annotations.On;
 import com.sap.cds.services.handler.annotations.ServiceName;
+import com.sap.cds.services.messages.Messages;
 import com.sap.cds.services.persistence.PersistenceService;
 
+import cds.gen.catalogservice.AddReviewContext;
 import cds.gen.catalogservice.Books;
+import cds.gen.catalogservice.Books_;
 import cds.gen.catalogservice.CatalogService_;
+import cds.gen.catalogservice.ReturnCatalogServiceSubmitOrder;
+import cds.gen.catalogservice.Reviews;
+import cds.gen.catalogservice.Reviews_;
+import cds.gen.catalogservice.SubmitOrderContext;
+import my.bookshop.BookRatingService;
+import my.bookshop.MessageKeys;
 
 /**
  * Custom business logic for the "Catalog Service" (see cat-service.cds)
  *
  * Handles Reading of Books
+ * 
  * Adds Discount Message to the Book Title if too much stock is available
+ * 
+ * Provides adding book reviews
  */
 @Component
 @ServiceName(CatalogService_.CDS_NAME)
 class CatalogServiceHandler implements EventHandler {
 
 	private final PersistenceService db;
 
-	CatalogServiceHandler(PersistenceService db) {
+	private final Messages messages;
+
+	private final BookRatingService bookRatingService;
+
+	private final CqnAnalyzer analyzer;
+	
+
+	CatalogServiceHandler(PersistenceService db, Messages messages, BookRatingService bookRatingService, CdsModel model) {
 		this.db = db;
+		this.messages = messages;
+		this.bookRatingService = bookRatingService;
+		this.analyzer = CqnAnalyzer.create(model);
+	}
+
+	/**
+	 * Invokes some validations before creating a review.
+	 * 
+	 * @param context {@link ReviewContext}
+	 */
+	@Before(entity = Books_.CDS_NAME)
+	public void beforeAddReview(AddReviewContext context) {
+		String reviewer = context.getUserInfo().getName();
+		String bookId = (String) analyzer.analyze(context.getCqn()).targetKeys().get(Books.ID);
+
+		Result result = db.run(Select.from(CatalogService_.REVIEWS)
+				.where(review -> review.book_ID().eq(bookId).and(review.reviewer().eq(reviewer))));
+
+		if (result.first().isPresent()) {
+			throw new ServiceException(ErrorStatuses.METHOD_NOT_ALLOWED,
+					"User not allowed to add more than one review for a given book").messageTarget(Reviews_.class,
+							r -> r.reviewer());
+		}
+	}
+
+	/**
+	 * Handles the review creation from the given context.
+	 * 
+	 * @param context {@link ReviewContext}
+	 */
+	@On(entity = Books_.CDS_NAME)
+	public void onAddReview(AddReviewContext context) {
+		Integer rating = context.getRating();
+		String title = context.getTitle();
+		String text = context.getText();
+
+		String username = context.getUserInfo().getName();
+		String bookId = (String) analyzer.analyze(context.getCqn()).targetKeys().get(Books.ID);
```

Comment:
In `validateBook` we anyway do a db request to check if the book exists. You can use the cqn statement from the context for that and from the resulting book you could easily get the book ID. Then you don't have to use cqn analyzer, because right now you extract the book ID to later construct the exact same statement that you already have here in the context.